test/test_util/test_data.py

```diff
@@ -0,0 +1,233 @@
+        f"sampled unique = {np.unique(y_sampled)}, original unique = {unique_labels}"
+
+
+@parametrize("X", [np.asarray([[1, 1, 1]] * 30)])
+@parametrize("x_type", [list, np.ndarray, csr_matrix, pd.DataFrame])
+@parametrize("y, task", [
+    (np.asarray([0] * 15 + [1] * 15), BINARY_CLASSIFICATION),
+    (np.asarray([0] * 10 + [1] * 10 + [2] * 10), MULTICLASS_CLASSIFICATION),
+    (np.asarray([[1, 0, 1]] * 30), MULTILABEL_CLASSIFICATION),
+    (np.asarray([1.0] * 30), REGRESSION),
+    (np.asarray([[1.0, 1.0, 1.0]] * 30), MULTIOUTPUT_REGRESSION),
+])
+@parametrize("y_type", [list, np.ndarray, pd.DataFrame, pd.Series])
+@parametrize("random_state", [0])
+@parametrize("sample_size", [0.25, 0.5, 5, 10])
+def test_subsample_validity(X, x_type, y, y_type, random_state, sample_size, task):
+    """ Asserts the validity of the function with all valid types """
+    assert len(X) == len(y)
+
+    if (
+        y_type == pd.Series
+        and task in [MULTILABEL_CLASSIFICATION, MULTIOUTPUT_REGRESSION]
+    ):
+        pytest.skip()
+
+    def convert(arr, objtype):
+        if objtype == np.ndarray:
+            return arr
+        elif objtype == list:
+            return arr.tolist()
+        else:
+            return objtype(arr)
+
+    X = convert(X, x_type)
+    y = convert(y, y_type)
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        X_sampled, y_sampled = subsample(
+            X, y,
+            random_state=random_state,
+            sample_size=sample_size,
+            is_classification=task in CLASSIFICATION_TASKS
+        )
+
+    # Check the types remain the same
+    def dtype(obj):
+        if isinstance(obj, list):
+            if isinstance(obj[0], list):
+                return type(obj[0][0])
+            else:
+                return type(obj[0])
+        elif isinstance(obj, pd.DataFrame):
+            return obj.dtypes
+        else:
+            return obj.dtype
+
+    if isinstance(X, pd.DataFrame):
+        assert list(dtype(X_sampled)) == list(dtype(X))
+    else:
+        print(X)
+        assert dtype(X_sampled) == dtype(X)
+
+    if isinstance(y, pd.DataFrame):
+        assert list(dtype(y_sampled)) == list(dtype(y))
+    else:
+        assert dtype(y_sampled) == dtype(y)
+
+    # Check the right amount of samples were taken
+    def size(obj):
+        if isinstance(obj, spmatrix):
+            return obj.shape[0] if obj.shape[0] > 1 else obj.shape[1]
+        else:
+            return len(obj)
+
+    # check the right amount of samples were taken
+    if sample_size < 1:
+        assert size(X_sampled) == int(sample_size * size(X))
+    else:
+        assert size(X_sampled) == sample_size
+
+
+@parametrize('X', [np.asarray([[0, 0, 1]] * 10)])
+@parametrize('dtype', supported_precision_reductions + [np.dtype('float32'), np.dtype('float64')])
+@parametrize('x_type', [np.ndarray, csr_matrix])
+def test_reduce_precision_correctly_reduces_precision(X, dtype, x_type):
```

Comment:
Same as below, can we have the expected outputs written down explicitly?

Reply:
Sure, I was using the mapping object but it's not explicitly written in the test which may be confusing. It reduce the coupling between the tests and the implementation (changing the mapping means going and changing the test) but it does make the test easier to read.

Reply:
Done